lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -72,44 +72,50 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
+  // This may be much larger than `log_2(cmod)`.
+  //
+  // Let `x` and `y` be the inputs to modular addition, then:
+  //    c1, n1 = addui_extended(x, y)
+  // If the coefficient modulus divides `2^N`, then return
+  //    c0 = c1 % cmod
+  // Otherwise, compute the adjusted result:
+  //    c0 = ((c1 % cmod) + (n1 * 2^N % cmod)) % cmod
   LogicalResult matchAndRewrite(
       AddOp op, OpAdaptor adaptor,
       ConversionPatternRewriter &rewriter) const override {
     auto type = adaptor.getLhs().getType();
-    auto elementTy =
-        type.cast<RankedTensorType>().getElementType().cast<IntegerType>();
-    auto extType = IntegerType::get(getContext(), elementTy.getWidth() + 1,
-                                    elementTy.getSignedness());
-    auto extShapeType = type.cast<ShapedType>().clone(extType);
-
-    mlir::Value extLhsOp = rewriter.create<arith::ExtUIOp>(
-        op.getLoc(), extShapeType, adaptor.getLhs());
-    mlir::Value extRhsOp = rewriter.create<arith::ExtUIOp>(
-        op.getLoc(), extShapeType, adaptor.getRhs());
-
-    auto addOp = rewriter.create<arith::AddIOp>(op.getLoc(), extShapeType,
-                                                extLhsOp, extRhsOp);
-
-    APInt mod =
-        op.getType().cast<PolynomialType>().getRing().coefficientModulus();
-    assert(mod != 0 && "coefficient modulus must not be zero");
-    auto modConstOp = rewriter.create<arith::ConstantOp>(
-        op.getLoc(), DenseElementsAttr::get(extShapeType, {mod}));
-
-    auto geCmp = rewriter.create<arith::CmpIOp>(
-        addOp.getLoc(), arith::CmpIPredicate::uge, addOp, modConstOp);
-    auto selectOp = rewriter.create<arith::SelectOp>(
-        geCmp->getLoc(), geCmp,
-        rewriter.create<arith::SubIOp>(geCmp->getLoc(), addOp, modConstOp),
-        addOp);
+    APInt mod = op.getType().cast<PolyType>().getRing().coefficientModulus();
+    auto cmod = rewriter.create<arith::ConstantOp>(
+        op.getLoc(), DenseElementsAttr::get(cast<ShapedType>(type), {mod}));
+
+    auto addExtendedOp = rewriter.create<arith::AddUIExtendedOp>(
+        op.getLoc(), adaptor.getLhs(), adaptor.getRhs());
+    auto c1ModOp = rewriter.create<arith::RemUIOp>(
+        op.getLoc(), addExtendedOp->getResult(0), cmod);
+    // If mod divides 2^N, c1modOp is our result.
+    if (mod.isPowerOf2()) {
+      rewriter.replaceOp(op, c1ModOp.getResult());
+      return success();
+    }
+    // Otherwise, add (n1 * 2^N % cmod)
+    APInt quotient, remainder;
+    APInt bigMod = APInt(mod.getBitWidth() + 1, 2) << (mod.getBitWidth() - 1);
+    APInt::udivrem(bigMod, mod.zext(bigMod.getBitWidth()), quotient, remainder);
+    remainder = remainder.trunc(mod.getBitWidth());
+
+    auto bitwidth = rewriter.create<arith::ConstantOp>(
+        op.getLoc(),
+        DenseElementsAttr::get(cast<ShapedType>(type), {remainder}));
+    auto adjustOp =
+        rewriter.create<arith::AddIOp>(op.getLoc(), c1ModOp, bitwidth);
```

Comment:
I wonder if this could also overflow...

Reply:
I don't think it can. I think I should write up a formal proof, so I'll file an issue on that.